bootloader/src/pyi_pythonlib.c

```diff
@@ -43,6 +43,107 @@
+ * If the archive is found, a pointer to the associated ARCHIVE_STATUS is returned
+ * otherwise the needed archive is opened and added to the pool and then returned.
+ * If an error occurs, returns NULL.
+ *
+ * Having several archives is useful for sharing binary dependencies with several
+ * executables (multipackage feature).
+ */
+static ARCHIVE_STATUS *
+_get_archive(ARCHIVE_STATUS *archive_pool[], const char *path)
+{
+    ARCHIVE_STATUS *archive = NULL;
+    int index = 0;
+    int SELF = 0;
+
+    VS("LOADER: Getting file from archive.\n");
+
+    for (index = 1; archive_pool[index] != NULL; index++) {
+        if (strcmp(archive_pool[index]->archivename, path) == 0) {
+            VS("LOADER: Archive found: %s\n", path);
+            return archive_pool[index];
+        }
+        VS("LOADER: Checking next archive in the list...\n");
+    }
+
+    archive = (ARCHIVE_STATUS *) calloc(1, sizeof(ARCHIVE_STATUS));
+
+    if (archive == NULL) {
+        FATALERROR("Error allocating memory for status\n");
+        return -1;
+    }
+
+    /*
+     * Setting this flag prevents creating another temp directory and
+     * the directory from the main archive status is used.
+     */
+    archive->has_temp_directory = archive_pool[SELF]->has_temp_directory;
+
+    if (pyi_arch_setup(archive, archive->homepath, &path[strlen(archive->homepath)])) {
+        FATALERROR("Error opening archive %s\n", path);
+
+        return -1;
+    }
+
+    archive_pool[index] = archive;
+    return archive;
+}
+
+/* Decide if the dependency identified by item is in a onedir or onfile archive
+ * then call the appropriate function.
+ */
+static int
+_extract_dependency(ARCHIVE_STATUS *archive_pool[], const char *item, const char* homepath)
+{
+    ARCHIVE_STATUS *status = NULL;
+    ARCHIVE_STATUS *archive_status = archive_pool[0];
+    char path[PATH_MAX];  /* The full path to the archive */
+    char filename[PATH_MAX];  /* The filename of the archive */
+    char srcpath[PATH_MAX];
+    char archive_path[PATH_MAX];  /* The filename within the archive */
+    char dirname[PATH_MAX];
+
+    VS("LOADER: Extracting dependencies\n");
+
+    pyi_path_basename(archive_path, item);
+    pyi_path_dirname(filename, item);
+    pyi_path_join(path, homepath, filename);
+
+    VS("LOADER: Attempting to extract %s from %s\n", archive_path, path);
+
+    /* Only extracting top-level dependencies from a onedir archive is currently supported */
+    status = _get_archive(archive_pool, path);
+
+    if (status == NULL) {
+        FATALERROR("Cannot open archive %s\n", filename);
+        return -1;
+    }
+
+    TOC * ptoc;
+
+    VS("LOADER: Searching for %s in archive\n", archive_path);
+
+    /* Iterate through toc looking for zlibs (PYZ, type 'z') */
+    ptoc = status->tocbuff;
+    while (ptoc < status->tocend) {
+        VS("LOADER: Archive item %s\n", ptoc->name);
+        if (ptoc->typcd == ARCHIVE_ITEM_PYZ) {
```

Comment:
This should be `&& strcmp(ptoc->name, archive_path)`, but I couldn't get that to work.